drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,313 @@
+				Direction:                network.SecurityRuleDirectionInbound,
+				Protocol:                 op.proto,
+			},
+		}
+	}
+	// Log open ports to be deleted and added
+	if len(openPortsToBeDeleted) > 0 || len(openPortsToBeAdded) > 0 {
+		removedPorts := make([]string, len(openPortsToBeDeleted))
+		addedPorts := make([]string, len(openPortsToBeAdded))
+		i := 0
+		for op := range openPortsToBeDeleted {
+			removedPorts[i] = op
+			i++
+		}
+		i = 0
+		for op := range openPortsToBeAdded {
+			addedPorts[i] = op
+			i++
+		}
+		log.Info("Updating security rule configuration", logutil.Fields{
+			"openPortsRemoved": removedPorts,
+			"openPortsAdded":   addedPorts,
+		})
+	}
+	// Remove all unnecessary open ports and release the priorities
+	for _, rule := range openPortsToBeDeleted {
+		delete(takenPriorities, to.Int32(rule.Priority))
+		removeSecurityRule(rules, rule)
+	}
+	// Add any necessary open ports
+	priority := int32(1000)
+	for _, rule := range openPortsToBeAdded {
+		// Find a priority that can be given to this rule
+		_, taken := takenPriorities[priority]
+		for taken {
+			priority++
+			_, taken = takenPriorities[priority]
+		}
+		rule.Priority = to.Int32Ptr(priority)
+		priority++
+		*rules = append(*rules, rule)
+	}
+	return nil
+}
+
+// checkIfDockerConflict does a rudimentary check to see if the rule could potentially conflict with the ability
+// for the docker port to be accessible. If the rule does not pass all rudimentary checks, it returns an error.
+func checkIfDockerConflict(rule network.SecurityRule, dockerPort int) error {
+	f := logutil.Fields{
+		"name":        to.String(rule.Name),
+		"description": to.String(rule.Description),
+		"priority":    to.Int32(rule.Priority),
+		"access":      rule.Access,
+		"direction":   rule.Direction,
+		"protocol":    rule.Protocol,
+		"portRanges":  getDestinationPorts(rule),
+	}
+	priority := to.Int32(rule.Priority)
+	if priority > dockerPriority {
+		return nil
+	}
+	purpose := strings.ToLower(fmt.Sprintf("%s %s %s", rule.Access, rule.Direction, rule.Protocol))
+	if purpose == "deny inbound tcp" || purpose == "deny inbound *" {
+		// check if docker port is covered by rule
+		for _, pr := range getDestinationPorts(rule) {
+			if checkIfPortInPortRange(dockerPort, pr) {
+				return fmt.Errorf("rule %s set to %s requests to the Docker port %d: %s",
+					to.String(rule.Name), purpose, dockerPort, f)
+			}
+		}
+	}
+	return nil
+}
+
+// checkIfPortInPortRange returns true if the port is within the portRange specified. If the port
+// range provided is improperly formatted, it returns false
+func checkIfPortInPortRange(port int, portRange string) bool {
+	if portRange == "*" {
+		return true
+	}
+	s := strings.Split(portRange, "-")
+	if len(s) == 2 {
+		// Port range represents is a range i.e. "2000-3000"
+		portMin, portMinErr := strconv.Atoi(s[0])
+		portMax, portMaxErr := strconv.Atoi(s[1])
+		if portMinErr == nil || portMaxErr == nil || portMin < portMax {
```

Comment:
should we check for errors from converting both bounds of the range? this way one could be invalid but considered as zero for this comparison `(port >= portMin) && (port <= portMax)`

Reply:
Seems like I flipped the Boolean logic here, should be:
```
if portMinErr == nil && portMaxErr == nil && portMin < portMax {
	return (port >= portMin) && (port <= portMax)
}
```
This change should check for errors from both bounds and, if an error occurs or the portMin / portMax is flipped, return false as expected.